### PR TITLE
opae-sdk: Set and Get  user clock using DFL UIO interface

### PR DESCRIPTION
**Pull request by anandaravuri**
Remove Iopll driver ( user clock driver ) and enable user clock DFL UIO  to access user clock  Device feature MMIO.

Move set/get user clock from Iopll driver to OPAE user space
Remove legacy/MCP user clock code

Signed-off-by: anandaravuri <ananda.ravuri@intel.com>

### CODE REVIEW

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 1]
```diff
@@ -0,0 +1,825 @@
...
+/* Field definitions for both USERCLKL_FREQ_CMD1 and USERCLK_FREQ_STS1 */
+#define IOPLL_CLK_MEASURE	    BIT_ULL(32)	/* Measure clk: 0=1x, 1=2x */
+#define IOPLL_FREQ_STS0		    0x18
+#define IOPLL_LOCKED		    BIT_ULL(60)
+#define IOPLL_AVMM_ERROR	    BIT_ULL(63)
+
+#define IOPLL_FREQ_STS1		    0x20
+#define IOPLL_FREQUENCY	        GENMASK_ULL(16, 0)	/* 10 kHz units */
+#define IOPLL_REF_FREQ	        GENMASK_ULL(50, 33)	/* Reference Freq */
+#define IOPLL_VERSION		    GENMASK_ULL(63, 60)	/* User clock version */
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:72]: Could we line up the spacing on these?

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+#define PLL_ENABLE_CALIBRATION	0x03
+
+#define IOPLL_MEASURE_LOW		0
+#define IOPLL_MEASURE_HIGH		1
+#define IOPLL_MEASURE_DELAY_MS	4000
+#define IOPLL_RESET_DELAY_MS	1000
+#define IOPLL_CAL_DELAY_MS		1000
+
+#define IOPLL_WRITE_POLL_INVL_US	10	/* Write poll interval */
+#define IOPLL_WRITE_POLL_TIMEOUT_US	1000000	/* Write poll timeout */
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:135]: Let's line up the spacing here, too.

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 3]
```diff
@@ -0,0 +1,825 @@
...
+
+
+
+fpga_result usrclk_reset(uint8_t* uio_ptr)
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:139]: We should turn these new files on in test-codingstyle-all.sh and fix any issues.

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+			}
+
+
+			}
+			break;
+		}
+
+
+free:
+	if (pglob.gl_pathv)
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:564]: Need to call opae_uio_close() here, else it it leaked.

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 5]
```diff
@@ -0,0 +1,825 @@
...
+	struct opae_uio uio;
+	uint8_t* uio_ptr = NULL;
+
+	if ((sysfs_path == NULL) ||
+		(userclk_high == NULL) ||
+		(userclk_low == NULL)) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:590]: You should memset &uio to 0 here.

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+	if (sysfs_path == NULL) {
+		OPAE_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	if (userclk_high < MIN_FPGA_FREQ) {
+		OPAE_ERR("Invalid Input frequency");
+		return FPGA_INVALID_PARAM;
+	}
+
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:664]: You should memset &uio to 0 here.

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 1 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c]: fixed

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 2 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c]: fixed

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 3 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c]: ran test-codingstyle-all.sh c and fixed all coding style issues

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 6 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c:658]: added memset , fixed

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 5 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c:585]: added memset , fixed

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 7]
```diff
@@ -0,0 +1,816 @@
...
+		*userclk_high = high * 1000;	// Adjust to Hz
+		*userclk_low = low * 1000;
+		return FPGA_OK;
+	} else if (ret == FPGA_NO_ACCESS) {
+		return FPGA_NO_ACCESS;
+	}
+
+	result = get_usrclk_uio(sysfs_path,
+		0x14,
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:602]: I think this is the feature id.  Can we use a #define for it?

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 8]
```diff
@@ -0,0 +1,816 @@
...
+		fd = open(sysfs_usrpath, O_WRONLY);
+		if (fd < 0) {
+			OPAE_MSG("open(%s) failed: %s",
+				sysfs_usrpath, strerror(errno));
+			return FPGA_NOT_FOUND;
+		}
+
+		cnt = sizeof(struct iopll_config);
+		do {
+			res = write(fd, bufp, cnt);
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:694]: I think write can be interrupted, returning EINTR.  I think Rodrigo has a routine for this.  Can we use it here?

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c [hunk 9]
```diff
@@ -0,0 +1,816 @@
...
+		return FPGA_NO_ACCESS;
+	}
+
+	struct pll_config *iopll_config = (struct pll_config *)bufp;
+	if ((iopll_config->pll_freq_khz > IOPLL_MAX_FREQ * 1000) ||
+		(iopll_config->pll_freq_khz < IOPLL_MIN_FREQ * 1000))
+		return FPGA_EXCEPTION;
+
+	result = get_usrclk_uio(sysfs_path,
+		0x14,
```
Comment (tswhison) [inline libraries/plugins/xfpga/usrclk/fpga_user_clk.c:715]: Can we use a #define for the feature id?

Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c]: fixed  added  #define USRCLK_FEATURE_ID 0x14

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 8 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c]: fixed  added  #define USRCLK_FEATURE_ID 0x14

File: libraries/plugins/xfpga/usrclk/fpga_user_clk.c (hunk 7 above)
Comment (anandaravuri) [reply libraries/plugins/xfpga/usrclk/fpga_user_clk.c]: fixed  added  #define USRCLK_FEATURE_ID 0x14

**Review by tswhison — APPROVED** [review]
Thanks!

**Review by michelleyho — APPROVED** [review]

**Pull request merged**